### PR TITLE
Add linters

### PR DESCRIPTION
**Pull request by abelsiqueira**
**Description**

Add linters. Originally I was planning to use [Schaapcommon](https://git.astron.nl/RD/schaapcommon), but they have a GPL-3.0 license, and we have Apache 2.0. Instead, I will use basic or default files.

**Related issues**:

- #23 

**Instructions to review the pull request**

Clone and verify that the linters run. There should be errors and modifications because I don't run or format anything on this PR.

Install [clang-format](https://clang.llvm.org/docs/ClangFormat.html), [clang-tidy](https://clang.llvm.org/extra/clang-tidy/), Python package [cmakelang](https://cmake-format.readthedocs.io/en/latest/installation.html), [cppcheck](https://cppcheck.sourceforge.io) and [flawfinder](https://dwheeler.com/flawfinder/)
```
cd $(mktemp -d --tmpdir cudawrappers-XXXXXX)
git clone https://github.com/nlesc-recruit/CUDA-wrappers .
git checkout <this-branch>
cmake -S . -B build
make -C build format
# compare to running individually
make -C build clang-format
make -C build cmake-format

make -C build lint
# compare to running individually
make -C build clang-tidy
make -C build cmake-lint
make -C build cppcheck
make -C build flawfinder
```

### CODE REVIEW

**Comment by jspaaks** [comment]
Hi Abel, nice work on this. What do you think about adding the install instructions for cmakelang, clang-tidy, clan-format to the developer documentation `README.dev.md` following Bouwe's suggestion?

**Comment by abelsiqueira** [comment]
> Hi Abel, nice work on this. What do you think about adding the install instructions for cmakelang, clang-tidy, clan-format to the developer documentation `README.dev.md` following Bouwe's suggestion?

Issue #32 touches that already.

**Comment by bouweandela** [comment]
Why do you need so many configuration files if you're using the defaults?

**Comment by abelsiqueira** [comment]
> Why do you need so many configuration files if you're using the defaults?

So the choices are explicit. Otherwise, the software will look for configuration files in whatever other places and that can lead to different results. The commands to generate these are in the files.

**Review by fdiblen — CHANGES_REQUESTED** [review]
thank you @abelsiqueira 

When I run `make -C build clang-lint` I get the error below.

```
make: Entering directory '/tmp/tmp.K30e8g7bzm/build'
make: *** No rule to make target 'clang-lint'.  Stop.
make: Leaving directory '/tmp/tmp.K30e8g7bzm/build'
```

Maybe we are missing a configuration for the linter?

**Comment by abelsiqueira** [comment]
Nice catch. Apparently, this is a bug with empty CMakeLists.txt in `usage-examples`. This can be verified with
```
cmake-lint usage-examples/plain-copy/CMakeLists.txt
```
While for the root CMakeLists.txt it will work normally. Notice that `cmake-format` fixes the empty CMakeLists.txt to have a single empty line, so running these in order, like I did, does not reveal this error.

**Comment by abelsiqueira** [comment]
> Including the generated files has the downside that changes to the default configuration will not be picked up as the linter/formatter is updated. Is that intentional? Or would it be better to e.g. just tell clang-format to use Google style instead of the Google style dumped to a configuration file?

Intentional. If I have a different version than you and we rely on the versions' default then we have different results. Also, I expected that ASTRON has some specific style, so this can also be seen as a placeholder for their choices.

File: .github/workflows/codacy-analysis.yml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+name: Codacy Analysis CLI
+
+on: ["push"]
+
+jobs:
+  codacy-analysis-cli:
```
Comment (fdiblen) [inline .github/workflows/codacy-analysis.yml:6]:
```suggestion
  codacy:
```

File: .github/workflows/codacy-analysis.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    name: Codacy Analysis CLI
```
Comment (fdiblen) [inline .github/workflows/codacy-analysis.yml:7]:
```suggestion
    name: Codacy
```

File: .github/workflows/codacy.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      - name: Run clang-tidy and codacy-clang-tidy
+        env:
+          CODACY_CLANG_TIDY_VERSION: 1.3.2
+          PROJECT_TOKEN: ${{ secrets.CODACY_PROJECT_TOKEN }}
+        run: |
+          export COMMIT=$(if [ ${{ github.event_name }} == "pull_request" ]; then echo "${{ github.event.pull_request.head.sha }}"; else echo "${{ github.sha }}"; fi)
+          wget https://github.com/codacy/codacy-clang-tidy/releases/download/${CODACY_CLANG_TIDY_VERSION}/codacy-clang-tidy-linux-${CODACY_CLANG_TIDY_VERSION} -O codacy-clang-tidy-latest
+          git clone https://github.com/codacy/codacy-clang-tidy
+          chmod a+x codacy-clang-tidy-latest codacy-clang-tidy/scripts/send-results.sh
+          cmake -S . -B build
```
Comment (abelsiqueira) [inline .github/workflows/codacy.yml:29]: These options do not have long versions

File: .github/workflows/codacy.yml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+
+      - name: Install requirements
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y clang-tidy nvidia-cuda-toolkit
+
+      # COMMIT variable defined below was taken from codacy-analysis-cli-action
```
Comment (abelsiqueira) [inline .github/workflows/codacy.yml:18]: Explaining where did `COMMIT` came from.

File: cmake/linter-tools.cmake [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+# Ref: https://stackoverflow.com/questions/32280717/cmake-clang-tidy-or-other-script-as-custom-target
+
+file(GLOB_RECURSE ALL_SOURCE_FILES *.cpp *.hpp)
+file(GLOB_RECURSE ALL_CMAKE_LISTS CMakeLists.txt)
+
+add_custom_target(
+    clang-format
+    COMMAND clang-format
+    -i
```
Comment (abelsiqueira) [inline cmake/linter-tools.cmake:12]: This does not have a long form.

File: cmake/linter-tools.cmake [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+)
+
+set(INCLUDE_DIRECTORIES -I../include/ -I${CUDAToolkit_INCLUDE_DIRS})
+add_custom_target(
+    clang-tidy
+    COMMAND clang-tidy
+    ${ALL_SOURCE_FILES}
+    --
+    -std=c++11
+    ${INCLUDE_DIRECTORIES}
```
Comment (abelsiqueira) [inline cmake/linter-tools.cmake:23]: `--` separates options for `clang-tidy` and options for the compiler.
Maybe `INCLUDE_DIRECTORIES` can be reused in the compilation of the rest of the library, but it is probably best discussed in a different issue.

**Comment by abelsiqueira** [comment]
@fdiblen, I looked into the `clang-tidy` variables, and it needs to build the library to work properly. That is why it needs `INCLUDE_DIRECTORIES`. However, I did not set INCLUDE_DIRECTORIES correctly.
At the moment, I am still not sure that Codacy + Clang-tidy is working properly, but I hope the other issues were addressed. Please let me know.

**Review by fdiblen — APPROVED** [review]
Looks good. Thanks @abelsiqueira :stew:

**Comment by abelsiqueira** [comment]
Thanks for the review :rice:

**Pull request merged**